include/aws/common/atomics.h

```diff
@@ -0,0 +1,247 @@
+ */
+
+enum aws_memory_order {
+    /**
+     * No particular ordering constraints are guaranteed relative to other
+     * operations at all; we merely ensure that the operation itself is atomic.
+     */
+    aws_memory_order_relaxed = 0,
+    /* aws_memory_order_consume - not currently implemented */
+
+    /**
+     * Specifies acquire ordering. No reads or writes on the current thread can be
+     * reordered to happen before this operation. This is typically paired with a release
+     * ordering; any writes that happened on the releasing operation will be visible
+     * after the paired acquire operation.
+     *
+     * Acquire ordering is only meaningful on load or load-store operations.
+     */
+    aws_memory_order_acquire = 2, /* leave a spot for consume if we ever add it */
+
+    /**
+     * Specifies release order. No reads or writes can be reordered to come after this
+     * operation. Typically paired with an acquire operation.
+     *
+     * Release ordering is only meaningful on store or load-store operations.
+     */
+    aws_memory_order_release,
+
+    /**
+     * Specifies acquire-release order; if this operation acts as a load, it acts as an
+     * acquire operation; if it acts as a store, it acts as a release operation; if it's
+     * a load-store, it does both.
+     */
+    aws_memory_order_acq_rel,
+
+    /*
+     * Specifies sequentially consistent order. This behaves as acq_rel, but in addition,
+     * all seq_cst operations appear to occur in some globally consistent order.
+     *
+     * TODO: Figure out how to correctly implement this in MSVC. It appears that interlocked
+     * functions provide only acq_rel ordering.
+     */
+    aws_memory_order_seq_cst
+};
+
+/* Include the backend implementation now, because we'll use its typedefs and #defines below */
+#if defined(__GNUC__) || defined(__clang__)
+
+#    include "atomics_gnu.inc"
+
+#elif defined(_MSC_VER)
+
+#    include "atomics_msvc.inc"
+
+#else
+
+#    error No atomics implementation for your compiler is available
+
+#endif
+
+/*
+ * Note: We do not use the C11 atomics API; this is because we want to make sure the representation
+ * (and behavior) of atomic values is consistent, regardless of what --std= flag you pass to your compiler.
+ * Since C11 atomics can silently introduce locks, we run the risk of creating such ABI inconsistencies
+ * if we decide based on compiler features which atomics API to use, and in practice we expect to have
+ * either the GNU or MSVC atomics anyway.
+ *
+ * As future work, we could test to see if the C11 atomics API on this platform behaves consistently
+ * with the other APIs and use it if it does.
+ */
+
+/**
+ * aws_atomic_int_t is defined to be an integer type that can fit into an struct aws_atomic_var.
+ * It is guaranteed to be at least 32 bits wide, but the size may vary depending on architecture.
+ */
+typedef aws_atomic_int_impl_t aws_atomic_int_t;
```

Comment:
is this going to be the same size as size_t ? could we make it use those semantics? can we make it unsigned?

Reply:
Hmm. I made this signed because I expect that use cases might want to add negative numbers, but if we assume twos complement that's not essential. As for size_t, we don't know for sure what sizes will be supported, but I suppose `size_t` is pretty likely to be supported, so I could maybe hard-code that.